src/tools/ceph_objectstore_tool.cc

```diff
@@ -680,55 +688,97 @@ int do_trim_pg_log(ObjectStore *store, const coll_t &coll,
 	continue;
       if (p->key().substr(0, 7) == string("missing"))
 	continue;
-      if (p->key().substr(0, 4) == string("dup_"))
-	continue;
-
       bufferlist bl = p->value();
       auto bp = bl.cbegin();
-      pg_log_entry_t e;
-      try {
-	e.decode_with_checksum(bp);
-      } catch (const buffer::error &e) {
-	cerr << "Error reading pg log entry: " << e.what() << std::endl;
-      }
-      if (debug) {
-	cerr << "read entry " << e << std::endl;
+
+      if (p->key().substr(0, 4) == string("dup_")) {
+        pg_log_dup_t dup;
+	try {
+	  dup.decode(bp);
+	  latest_dups.push_back(dup);
+	  if (latest_dups.size() > tracked_dups) {
+	    pg_log.dups.push_back(latest_dups.front());
+	    latest_dups.pop_front();
+	  }
+	  if (debug) {
+            cerr << "read entry " << dup << std::endl;
+          }
+	} catch (buffer::error& e) {
+	  cerr << "error reading log dup: " << p->key() << std::endl;
+	  return -EFAULT;
+	}
+      } else {
+        pg_log_entry_t e;
+        try {
+	  e.decode_with_checksum(bp);
+	  if (e.version > trim_to) {
+	    continue;
+	  }
```

Comment:
That looks weird. It seems the idea was to use `pg_log.trim` in order not to do the trimming manually, but we still need this.

Reply:
i thought we can stop collecting entries, but yes, we can let trim function do it for us